File: scripts/summarize-md.ts
import { OpenAI } from "langchain/llms/openai";
import { loadSummarizationChain, AnalyzeDocumentChain } from "langchain/chains";
import { RecursiveCharacterTextSplitter } from "langchain/text_splitter";
import { GithubRepoLoader } from "langchain/document_loaders/web/github";
import { readFileSync, writeFile, writeFileSync } from "fs";
import { CheerioWebBaseLoader } from "langchain/document_loaders/web/cheerio";

/*
  1. Ingest a markdown file and extract each link.
  2. For each link, read each webpage and summarize it.
  3. Create sumaries at multiple skill levels for each link ("Detailed", "ELI5")
  4. Create a new markdown file with the summaries.
*/
const NYC_BITDEVS = 'https://github.com/BitDevsNYC/BitDevsNYC.github.io';

const loadFromRepo = async () => {
  const loader = new GithubRepoLoader(
    NYC_BITDEVS,
    { branch: "main", recursive: false, unknown: "warn", ignorePaths: ["*.md"] }
  );
  const docs = await loader.load();
};

type Link = {
  full: string;
  title: string;
  link: string;
  summary?: string;
}


/**
 * Ingest a markdown file and extract each link.
 */
const getLinks = async (markdownPath: string): Promise<Link[]> => {
  const regexMdLinks = /\[([^\[]+)\](\(.*\))/gm // gets all markdown links 
  const singleMatch = /\[([^\[]+)\]\((.*)\)/ // gets the url
  const file = await readFileSync(markdownPath, "utf-8");
  const matches = file.match(regexMdLinks) ?? [];
  const links = []
  for (let i = 0; i < matches.length; i++) {
    const match = singleMatch.exec(matches[i])
    // const match = matches[i];
    if (!match) continue;
    links.push({
      full: match[0],
      title: match[1],
      link: match[2]
    })
  }
  return links;
}
async function summarizeLinks(links: Link[]): Promise<Link[]> {
  const result = [];
  for (let i = 0; i < links.length; i++) {
    const link = links[i];
    const summary = await summarizeWebpage(link.link).catch(e => console.error(e));
    if (!summary) continue;
    console.log(summary);
    result.push({...link, summary: summary.text});
  }
  return result

}

export async function summarizeWebpage(link: string) { 
  if (!link) return;
  const model = new OpenAI({ temperature: 0, maxTokens: 2000  });
  const textSplitter = new RecursiveCharacterTextSplitter({ chunkSize: 4000 });

  const loader = new CheerioWebBaseLoader(link);
  
  const docs = await loader.loadAndSplit(textSplitter);
  console.log('docs length', docs.length);
  const combineDocsChain = await loadSummarizationChain(model, { type: "map_reduce" });
  try {
    const summary = await combineDocsChain.call({input_documents: docs}) 
    return summary;
  } catch (e) {
    // @ts-ignore 
    console.error(e?.response?.data);
    return null;
  }
}

const writeSummariesToFile = async (filename: string, summaries: Link[]) => {
  const pathPrefix = './_summaries/';
  const fullPath = pathPrefix + filename;
  const titleLine = '# ' + filename + ' Link Summaries\n\n'
  writeFileSync(fullPath, titleLine);
  const result = []
  for (let i = 0; i < summaries.length; i++) {
    //  * [Title](link)
    //     * Summary
    const lineToWrite = '* [' + summaries[i].title + '](' + summaries[i].link + ')\n    * ' + summaries[i].summary + '\n\n';
    result.push(lineToWrite);
  }
  writeFileSync(fullPath, result.join(''))

};

const run = async () => { 
  const fileName = '2023-06-28-socratic-seminar-125.md';
  const pathPrefix = './_posts/';
  const links = await getLinks(pathPrefix + fileName);
  const summaries = await summarizeLinks(links);
  console.log('# of summaries', summaries);
  const result = await writeSummariesToFile('2023-06-28-socratic-seminar-125.md', summaries);
};
run()

File: scripts/py/summarize-md.py
import asyncio
import re
from typing import List, Dict, Optional
import requests
from bs4 import BeautifulSoup
from openai import OpenAI, ChatCompletion

# Equivalent to OpenAI in Node.js, note that you'll need to obtain an API key from OpenAI and set it as an environment variable
openai = OpenAI("your_api_key_here")

def load_from_repo():
    # Placeholder for code to load a repo using GitHub's API
    pass

def get_links(markdown_path: str) -> List[Dict[str, str]]:
    regex_md_links = re.compile(r"\[([^\[]+)\]\((.*)\)") 
    with open(markdown_path, "r") as file:
        content = file.read()
    matches = re.findall(regex_md_links, content)
    links = [{'full': f"[{m[0]}]({m[1]})", 'title': m[0], 'link': m[1]} for m in matches]
    return links

async def summarize_links(links: List[Dict[str, str]]) -> List[Dict[str, str]]:
    tasks = [summarize_webpage(link) for link in links]
    results = await asyncio.gather(*tasks, return_exceptions=True)
    summaries = [res for res in results if not isinstance(res, Exception)]
    return summaries

async def summarize_webpage(link: Dict[str, str]) -> Optional[Dict[str, str]]:
    if not link:
        return None
    response = requests.get(link['link'])
    soup = BeautifulSoup(response.text, 'html.parser')
    text = ' '.join(soup.stripped_strings)
    chat_model = 'gpt-4.0-turbo'  # Assuming the usage of gpt-4.0-turbo as the model
    message = [{'role': 'system', 'content': "You are a helpful assistant."}, {'role': 'user', 'content': f"Please summarize the following text: {text}"}]
    response = openai.ChatCompletion.create(model=chat_model, messages=message)
    if response['choices'][0]['finish_reason'] == 'stop':
        return {**link, 'summary': response['choices'][0]['message']['content']}
    return None

def write_summaries_to_file(filename: str, summaries: List[Dict[str, str]]):
    path_prefix = './_summaries/'
    full_path = path_prefix + filename
    title_line = f'# {filename} Link Summaries\n\n'
    with open(full_path, "w") as file:
        file.write(title_line)
        for summary in summaries:
            line_to_write = f"* [{summary['title']}]({summary['link']})\n    * {summary['summary']}\n"
            file.write(line_to_write)

async def run():
    file_name = '2023-06-28-socratic-seminar-125.md'
    path_prefix = './_posts/'
    links = get_links(path_prefix + file_name)
    summaries = await summarize_links(links[:10])
    print(f'# of summaries {len(summaries)}')
    write_summaries_to_file('2023-06-28-socratic-seminar-125.md', summaries)

# This is required to run asyncio code in Python
asyncio.run(run())
